actors/builtin/miner/miner_actor.go

```diff
@@ -151,42 +157,71 @@ func (a Actor) ControlAddresses(rt Runtime, _ *adt.EmptyValue) *GetControlAddres
 	rt.State().Readonly(&st)
 	info := getMinerInfo(rt, &st)
 	return &GetControlAddressesReturn{
-		Owner:  info.Owner,
-		Worker: info.Worker,
+		Owner:        info.Owner,
+		Worker:       info.Worker,
+		ControlAddrs: info.ControlAddresses,
 	}
 }
 
 type ChangeWorkerAddressParams struct {
-	NewWorker addr.Address
+	NewWorker       addr.Address
+	NewControlAddrs []addr.Address
 }
 
 func (a Actor) ChangeWorkerAddress(rt Runtime, params *ChangeWorkerAddressParams) *adt.EmptyValue {
 	var effectiveEpoch abi.ChainEpoch
-	worker := resolveWorkerAddress(rt, params.NewWorker)
+
+	worker := addr.Undef
+	if params.NewWorker != addr.Undef {
+		worker = resolveWorkerAddress(rt, params.NewWorker)
+	}
+
+	controlAddrs := make([]addr.Address, 0, len(params.NewControlAddrs))
+	for _, ca := range params.NewControlAddrs {
+		resolved := resolveControlAddress(rt, ca)
+		controlAddrs = append(controlAddrs, resolved)
+	}
+
 	var st State
 	rt.State().Transaction(&st, func() {
 		info := getMinerInfo(rt, &st)
 
-		// Only the Owner is allowed to change the worker address.
-		// Allowing the worker to do this does not make sense because the owner would usually do it
-		// only if the worker keys are lost.
+		// Only the Owner is allowed to change the worker and control addresses.
 		rt.ValidateImmediateCallerIs(info.Owner)
 
-		effectiveEpoch = rt.CurrEpoch() + WorkerKeyChangeDelay
+		{
+			// save the new control addresses
+			if len(controlAddrs) != 0 {
```

Comment:
The miner must be able to remove all these addresses. Remove this check, just replace the addresses with whatever was provided.

Reply:
This is done.